teps/0058-graceful-pipeline-run-termination.md

```diff
@@ -109,8 +107,23 @@ spec:
   status: "PipelineRunStopped"
 ```
 
--->
+In the second scenario, users want to wait for running tasks to be completed.
+To gracefully terminate a `PipelineRun`, users update its definition to mark it as stopping. 
+When you do so, the spawned `TaskRuns` are not cancelled.
+The final tasks are triggered, when all running tasks are finalized.
```

Comment:
Hmm this second scenario is expanding the scope of the proposal somewhat from the original problem statement

1. I'm not sure that the difference between `PipelineRunStopped` and `PipelineRunStopping` is enough to convey this difference (I had to re-read this several times before I realized this was proposing TWO new states, I thought one was a typo.... 😅 )
2. I don't fully understand the use cases for the new scenario, i.e. the one where you want to cancel execution but not stop any already running TaskRuns, is it possible to elaborate on this in the use cases section?

Reply:
`PipelineRunStopping` is a status we use to track the fact that now new task runs will be scheduled because of failure of some kind. For the user we could have different names:
- `PipelineRunCancelled`: forced cancel, same as today
- `PipelineRunCancelledGracefully`: graceful cancel, finally run
- `PipelineRunCancelledReallyGracefully` (not a serious name, any good ideas?): wait for running tasks to stop, do no schedule any new task, then run finally

Reply:
@bobcatfish the new case allows to wait for already running TaskRuns to produce expected results, but prevents others from being triggered. This is quite common case, when you realise that further processing is worthless (e.g. some conditions have changes), but the intermediate result is valuable even for investigation.

Additionally, in Kubeflow Pipelines with results cache being enabled, you can rerun pipeline reusing intermediate results. That combined with "super-graceful" termination allows you to continue processing from the place, where it was stopped.

Reply:
I don't have any good example for option 4. in mind - wait for running but skip finally, which is expected to perform cleanup.

If the system is not fully blocked, by some unexpected behaviour, I wouldn't cancel a PipelineRun forcefully and expose myself for potential resource leaks.